lib/srv/sessiontracker_test.go

```diff
@@ -27,8 +27,128 @@ import (
+	services.SessionTrackerService
+	updated     chan struct{}
+	updateError chan error
+}
+
+func (f *failingTrackerService) CreateSessionTracker(ctx context.Context, s types.SessionTracker) (types.SessionTracker, error) {
+	return s, nil
+}
+
+func (f *failingTrackerService) UpdateSessionTracker(ctx context.Context, req *proto.UpdateSessionTrackerRequest) error {
+	f.updated <- struct{}{}
+
+	return <-f.updateError
+}
+
+func TestSessionTracker_UpdateRetry(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	clock := clockwork.NewFakeClock()
+
+	svc := &failingTrackerService{
+		updated:     make(chan struct{}),
+		updateError: make(chan error),
+	}
+	spec := types.SessionTrackerSpecV1{
+		Created:   clock.Now(),
+		SessionID: "session",
+		Expires:   clock.Now().Add(10 * time.Hour),
+	}
+	tracker, err := NewSessionTracker(ctx, spec, svc)
+	require.NoError(t, err)
+	done := make(chan error)
+
+	updateError := trace.ConnectionProblem(context.DeadlineExceeded, "connection refused")
+
+	// start the update loop
+	go func() {
+		done <- tracker.updateExpirationLoop(ctx, clock)
+	}()
+
+	for i := 0; i < 10; i++ {
+		// wait for the ticker to be ready
+		if i%2 != 0 {
+			clock.BlockUntil(2)
+		} else {
+			clock.BlockUntil(1)
+		}
+
+		// advance the clock to fire the ticker
+		clock.Advance(sessionTrackerExpirationUpdateInterval)
+
+		// wait for update to be called
+		select {
+		case <-svc.updated:
+		case err := <-done:
+			t.Fatal("Update loop terminated early", err.Error())
+		}
+
+		// send back an error on even iterations
+		if i%2 == 0 {
+			svc.updateError <- updateError
+
+			// wait for the retry to be engaged
+			clock.BlockUntil(1)
+
+			// advance far enough for the retry to fire
+			clock.Advance(45 * time.Second)
+
+			// wait for the update to be called again
+			select {
+			case <-svc.updated:
+			case err := <-done:
+				t.Fatal("Update loop terminated early", err.Error())
+			}
+		}
+
+		svc.updateError <- nil
+	}
+
+	// advance the clock for one last update attempt
+	clock.BlockUntil(1)
+	clock.Advance(sessionTrackerExpirationUpdateInterval)
+
+	// wait for update to be called and return an error to
+	// get in the retry path
+	select {
+	case <-svc.updated:
+		svc.updateError <- updateError
+	case err := <-done:
+		t.Fatal("Update loop terminated early", err.Error())
+	}
+
+	// advance far enough for the retry to fire
+	clock.BlockUntil(1)
+	clock.Advance(45 * time.Second)
+
+	// wait for update to be called from the retry loop and return an error
+	select {
+	case <-svc.updated:
+		// advance the clock again to make the tracker be stale
+		// enough to abort the retry loop
+		clock.Advance(10 * time.Hour)
+		svc.updateError <- updateError
+	case err := <-done:
+		t.Fatal("Update loop terminated early", err.Error())
+	}
+
+	// ensure the update loop ends
+	select {
+	case err := <-done:
+		require.ErrorIs(t, err, updateError)
+	case <-time.After(10 * time.Second):
```

Comment:
can this be shortened? if this does not complete quickly I can't see how it would be different after 10 seconds - I know test runners can be slow sometimes, bit this slow? Unnecessarily slow tests can be rather annoying.